### ISSUE TITLE
Make it possible to reason about content to project (content projection and content queries unification)

### ISSUE THREAD

**Issue by pkozlowski-opensource** [angular/angular#37319]
There are multiple use-cases where one would like to reason about content provided to a component  (check if content was provided at all, count the number of items etc.). 

To make the discussion more concrete let's consider the following markup ([full runnable scenario](https://stackblitz.com/edit/angular-ivy-sumkgl?file=src%2Fapp%2Fapp.component.ts)):

```html
<my-container>
      <my-item></my-item>
      <my-item></my-item> 
      <my-item></my-item>
</my-container>
```

Given this markup the `<my-container>` component might want to answer the following questions:
* was content provided at all (existing issue.: )
* how many items were provided (as of today this could be done through content queries only).

On top of this the `<my-container>` implementation might want to do the following (all the provided examples boil down to the ability of projecting items individually):
* wrap each item into some markup;
* project only 2 first items;
* project every second item (or items in 2 separate columns);
* etc. etc.

Achieving any of the use-cases listed above is very difficult in the current Angular and it steams from the fact that one can't reason (count, pick-up, check presence etc.) about individual items to project. 

While content queries can be used, to some degree, to reason about content items individually, the content queries approach has its problems as well:
* we end up with 2 mechanism operating on content (content queries and content projection) - those 2 mechanism have different APIs, different capabilities, need to be learnt separately etc.;
* there is a fundamental mismatch between 2 mechanism: content projection operates on "blocks" that can't be reason about (but capture content), while content queries are good with individual items but don't capture content (we can't insert result of a content query into the DOM).

This fundamental mismatch between content projection and content queries leads to many surprising situations and use-cases that can't be implemented really. Again, here is a "typical" example: https://stackblitz.com/edit/angular-ivy-sumkgl?file=src%2Fapp%2Fapp.component.ts

In short: content projection and content queries are good mechanisms while used separately, but as soon as we try to use both mechanisms together we start to bump into serious problems.

The idea behind opening this issue is to capture use-cases that are difficult to implement and expose problems where both content queries and content projection are used together. This should open ways to designing a more powerful content projection where it is possible to reason about and project content with one API. To set expectations: there are no immediate implementation plans, we want to capture use-cases and experiment with different APIs first.

**Comment by bmtheo** [comment]
Here is one use case with the Angular Material lib : (ref for more info : https://github.com/angular/components/issues/9411)

The use case is to share some advanced form fields in an app (add features around a MatFormFieldControl) without reimplementing all the values of input

The content projection is used to project the <input> into the advanced form field
The content queries are used by Angular Material to manipulate the input

Rough example : 
```
<!-- my-advanced-field.component.html -->
<mat-form-field>
  <ng-content></ng-content>
  <!-- some other features of my advanced field -->
</mat-form-field>

<!-- my-component.component.html -->
<my-advanced-field>
  <input matInput ... >
</my-advanced-field>
```
Full example here : https://stackblitz.com/edit/angular-material2-issue-jekywv

Without this, all the values of the input used by my-component needs to be implemented by my-advanced-field resulting in large boilerplate code

**Comment by NachoBoss** [comment]
any update? Bump.

**Comment by mtuzinskiy** [comment]
Any updates on this?  
Here's a bright issue: 
```
<mat-form-field>
  <ng-content></ng-content>
  <mat-error>My reusable error</mat-error>
</mat-form-field>
```
...aand you cannot do it.  And it's 2022.

**Comment by MilesAl** [comment]
This "feature" should be implemented as soon as possible. There are some use cases that have this issue and the larger the application scales, the more issues arise. The mat-menu becomes partially unusable or requires a lot of redundant code, which is not acceptable for enterprise applications. Please work on this issue!

**Comment by matheo** [comment]
+1 here
I'm trying to programatically create a control dynamically with a wrapper
and `@ContentChild` cannot fetch the projected component:
![image](https://user-images.githubusercontent.com/260185/183141604-877b685f-d660-40f6-b4ee-7be904d781bd.png)

dyn-factory.html:
```html
<ng-container #control></ng-container>
<ng-container #wrapper></ng-container>
```

dyn-factory.ts:
```typescript
this.controlRef = this.controlContainer.createComponent<AbstractDynControl>(controlFactory, undefined, injector);

this.wrapperRef = this.wrapperContainer.createComponent<AbstractDynWrapper>(
  wrapperFactory,
  undefined,
  injector,
  [[this.controlRef.location.nativeElement]],
);
```

and having this simple wrapper:
```html
<p>form-field</p>
<ng-content></ng-content>
```
for this simple control:
```html
<input matInput [type]="params.type" [formControl]="control" />
```
works in the HTML output but the wrapper is not able to fetch the `WrapperComponent` nor the `MatFormFieldControl`
via `@ContentChild` nor `@ContentChildren(x, { descendants: true })`

help with this feature please :(

**Comment by csisy** [comment]
> Any updates on this? Here's a bright issue:
> 
> ```
> <mat-form-field>
>   <ng-content></ng-content>
>   <mat-error>My reusable error</mat-error>
> </mat-form-field>
> ```
> 
> ...aand you cannot do it. And it's 2022.

Yep. We wanted the same thing and then found this issue which is open since May 28 2020...

**Comment by lcervo-diligent** [comment]
any updates?

**Comment by gegis** [comment]
Up+

**Comment by BenRacicot** [comment]
I have to be that guy 😂 first to bump this in 2023!

**Comment by SatarisGIT** [comment]
+1

**Comment by SatarisGIT** [comment]
... and next bump :P

**Comment by haskelcurry** [comment]
They are too busy with Signals or other useless crap, guys

**Comment by n614cd** [comment]
I also ran into one of the use cases listed above. That is with the Angular material components. The only workaround ends up being the replication of extensive amounts of boiler plate code in HTML.

**Comment by j-a-m-l** [comment]
I just want to point out that maybe some of you could try https://github.com/angular/components/issues/9411#issuecomment-596075768. 

I'm grateful to @GrandSchtroumpf for its workaround.

**Comment by csisy-bt4w** [comment]
Another stackblitz for a simple use-case, where the content of a child component is queried via a directive but the content is provided through a parent: https://stackblitz.com/edit/stackblitz-starters-fndspl?file=src%2Fmain.ts

**Comment by ngaritagoitia** [comment]
bump again

**Comment by vlio20** [comment]
Here is another example:
```
<mat-menu #menu>
  <ng-content select="[mat-menu-item]"></ng-content>
  <button mat-menu-item>Logout</button>
</mat-menu>
```

**Comment by tmb-jyskefinans-dk** [comment]
> > Any updates on this? Here's a bright issue:
> > ```
> > <mat-form-field>
> >   <ng-content></ng-content>
> >   <mat-error>My reusable error</mat-error>
> > </mat-form-field>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ...aand you cannot do it. And it's 2022.
> 
> Yep. We wanted the same thing and then found this issue which is open since May 28 2020...

Really hacky way could be to do as this for now:
```
<ng-content ngProjectAs="mat-error">
</ng-content>
<mat-error></mat-error>
  ```
  and in your style display:none the last-of-type of the mat-error tricks the component to handle the mat-error correct but hides the dummy mat-error that would normally take up some space.

**Comment by BenRacicot** [comment]
Do we still have this problem within control flow?

```
@if (true) {
  <div></div> 
}
```

**Comment by Fasteroid** [comment]
To quote the developers of TF2, 
_“`// this is utterly fucking retarded`”_

WHY can we not do this yet?  Bump 2025 smh my head

**Comment by Fasteroid** [comment]
> > > Any updates on this? Here's a bright issue:
> > > ```
> > > <mat-form-field>
> > >   <ng-content></ng-content>
> > >   <mat-error>My reusable error</mat-error>
> > > </mat-form-field>
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ...aand you cannot do it. And it's 2022.
> > 
> > 
> > Yep. We wanted the same thing and then found this issue which is open since May 28 2020...
> 
> Really hacky way could be to do as this for now:
> 
> ```
> <mat-form-field appearance="outline">
>   <input matInput type="text" />
>   <ng-content ngProjectAs="mat-error">
>   </ng-content>
>   <mat-error></mat-error>
> </mat-form-field>
> ```
> 
> css
> 
> ```
> mat-error:last-of-type {
>      display:none;
> }
> ```
> 
> Then in when using your wrapper component you can do:
> 
> ```
> <custom-input>
>      <mat-error>Fix</mat-error>
> </custom-input>
> ```
> 
> This tricks the component to handle the mat-error correct but hides the dummy mat-error that would normally take up some space.

Great hack, is there any approach that works with multiple projected elements?  I can't figure out why yours works so well.

**Comment by mrmokwa** [comment]
The fallback to default ng-content when empty ng-content is avaiable since v19 was released.
We should check if this PR still necessary.

Commit: https://github.com/angular/angular/commit/7b1e5be20b99c88246c6be78a4dcd64eb55cee1a
